mlir/lib/Quantum/Transforms/MergeRotationsPatterns.cpp

```diff
@@ -49,27 +52,105 @@ struct MergeRotationsRewritePattern : public mlir::OpRewritePattern<CustomOp> {
             return failure();
         }
 
-        TypeRange outQubitsTypes = op.getOutQubits().getTypes();
-        TypeRange outQubitsCtrlTypes = op.getOutCtrlQubits().getTypes();
-        ValueRange parentInQubits = parentOp.getInQubits();
-        ValueRange parentInCtrlQubits = parentOp.getInCtrlQubits();
-        ValueRange parentInCtrlValues = parentOp.getInCtrlValues();
-
-        auto parentParams = parentOp.getParams();
-        auto params = op.getParams();
-        SmallVector<mlir::Value> sumParams;
-        for (auto [param, parentParam] : llvm::zip(params, parentParams)) {
-            mlir::Value sumParam =
-                rewriter.create<arith::AddFOp>(loc, parentParam, param).getResult();
-            sumParams.push_back(sumParam);
-        };
-        auto mergeOp = rewriter.create<CustomOp>(loc, outQubitsTypes, outQubitsCtrlTypes, sumParams,
-                                                 parentInQubits, opGateName, nullptr,
-                                                 parentInCtrlQubits, parentInCtrlValues);
-
-        op.replaceAllUsesWith(mergeOp);
-
-        return success();
+        if (opGateName == "qml.Rot" || opGateName == "qml.CRot") {
```

Comment:
What happens if the rotation gates are adjointed? Should the merge still happen?

(In Catalyst adjointed gates are indicated by a adjoint unit attribute, see for example [here](https://github.com/PennyLaneAI/catalyst/blob/main/mlir/test/Quantum/ChainedSelfInverseTest.mlir#L421)

Reply:
Thank you @paul0403 

Please let me know if you did not consider whether the rotation gates are adjointed for regular merging rotations (not for merging non-commutative rotations), or if we do not have adjointed gates for regular merging rotations?

Reply:
The merge rotation pass applies an [rotation gate adjoint canonicalization](https://github.com/PennyLaneAI/catalyst/blob/main/mlir/lib/Quantum/Transforms/merge_rotation.cpp#L64). The canonicalization simply changes all angles to their negative and removes the adjoint attribute. See https://github.com/PennyLaneAI/catalyst/pull/1205

However, looking at the canonicalization pattern, you will find that `Rot` and `CRot` are not canonicalized.
(a) Why do you think that is?
(b) Knowing this, what do you think you should do in your added pattern here (assuming no new canonicalization is added for `(C)Rot`)?

Reply:
Thank you @paul0403 
(a) Rot and CRot are not canonicalized because they involve complex, multi-parameter rotations that cannot be standardized by simply negating a single parameter. 
(b) First, check for the adjoint attribute on qml.Rot and qml.CRot. If the operation is adjointed, transform it into its non-adjointed, canonical form by reversing the order of the parameters and negating each parameter. After this transformation, remove the adjoint attribute to standardize the operation. Then, proceed with the merging process as if all rotations are in canonical form, ensuring consistency across operations.
Please let me know if my method is correct.

Reply:
Good insights! This is what I would do as well. 

Due to how the work is organized (aka adjoint canonicalization happens somewhere else, not here in the merge rotation patterns), in the merge rotation patterns, it suffices to assume that the rotation gates will not carry adjoint attributes when the patterns are hit.

Thus the only thing needed here is a check that the `(C)Rot` gates do not carry adjoint attributes. If they do, the pattern should do nothing.

Reply:
Thank you, added it.